[libc][OSUtil] refactor quick_exit to be an object library everywhere

nickdesaulniers opened the pull request:
The usage of __builtin_unreachable after calls to quick_exit were distressing.
If a function is properly marked [[noreturn]] then __builtin_unreachable is not
necessary.

Looking into this further, we seem to have header only implementations for CPU
targets. The inline nature of these functions is curious; we're going to exit,
it doesn't matter if we need to pay the call of a function or not. If we just
make these functions have distinct TUs rather than be header only, we can clean
up the cmake rules for quick_exit which were different between CPU and GPU.

Remove darwin support for quick_exit. This isn't being tested, and we can bring
it back when necessary.

nickdesaulniers commented on libc/src/__support/OSUtil/linux/quick_exit.cpp:
```diff
@@ -0,0 +1,15 @@
+
```
ah, missing header block. Let me go back and `git mv` the .h to this .cpp file, then strip out the header guards.

jhuber6 commented on libc/src/__support/OSUtil/gpu/quick_exit.cpp:
```diff
@@ -6,17 +6,14 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef LLVM_LIBC_SRC___SUPPORT_OSUTIL_GPU_QUICK_EXIT_H
-#define LLVM_LIBC_SRC___SUPPORT_OSUTIL_GPU_QUICK_EXIT_H
```
Wonder how these got here, probably some victim of copy paste.

nickdesaulniers replied:
it happens

jhuber6 reviewed (approved):
I think this is reasonable. Ideally the targets all depend on `osutil` correctly so we'll pull this in where needed.

SchrodingerZhu reviewed (approved):
I second the changes here.

jhuber6 commented on libc/src/__support/OSUtil/linux/quick_exit.cpp:
```diff
@@ -22,14 +17,12 @@ namespace LIBC_NAMESPACE {
 #ifdef LIBC_TARGET_ARCH_IS_X86
 __attribute__((no_stack_protector))
 #endif
-LIBC_INLINE void
-quick_exit(int status) {
+__attribute__((noreturn))
+void quick_exit(int status) {
```
Any way we could use C++ `[[]]` syntax for these?

SchrodingerZhu replied:
Actually, `[[gnu::no_stack_protector]]` was added to clang because of this.

jhuber6 replied:
I see, adding attributes is fairly trivial, but unlike the GPU build we can't rely on an up-to-date compiler so I guess we'll just need to keep it as-is.

nickdesaulniers replied:
yeah, I hit an issue with having these be repeated as separate attribute expressions, but I think I know how to fix that.

nickdesaulniers replied:
done in https://github.com/llvm/llvm-project/pull/85955/commits/9ceeba4d5d7db1d58eeea33c31e46e9c5aea60a1 PTAL

nickdesaulniers replied:
the issue was, clang errors if you mix GNU C `__attribute__`s with C23/C++20 `[[]]` attributes.

nickdesaulniers replied:
[presubmit failure](https://buildkite.com/llvm-project/github-pull-requests/builds/49514#018e62f3-9c12-4483-9ee1-1f37a522324e)

> -- The C compiler identification is Clang 16.0.6
> llvm-project/github-pull-requests/libc/src/__support/OSUtil/linux/quick_exit.cpp:18:3: error: unknown attribute 'no_stack_protector' ignored [-Werror,-Wunknown-attributes]
[[gnu::no_stack_protector]]
  ^~~~~~~~~~~~~~~~~~~~~~~

So ironically, `no_stack_protector` was a function attribute in clang first, then GCC added support, then clang had to match the `gnu::` prefix later.  GCC does not recognize the `clang::` attribute prefix.  All the issues with compiler version compatibility are avoided by just using the GNU C extension syntax for attributes.

So I'm going to revert https://github.com/llvm/llvm-project/pull/85955/commits/9ceeba4d5d7db1d58eeea33c31e46e9c5aea60a1.

jhuber6 replied:
Yeah, probably should've been more clear. This will work fine if you require runtimes builds of `clang` since it has the change, but not with the minimum clang version we support.

nickdesaulniers commented:
TODO: breaks bazel!

michaelrj-google reviewed (approved):
LGTM once bazel is fixed.

nickdesaulniers commented on libc/src/__support/OSUtil/baremetal/quick_exit.h:
```diff
@@ -1,21 +0,0 @@
-namespace LIBC_NAMESPACE {
-
-// This is intended to be provided by the vendor.
-extern "C" void __llvm_libc_quick_exit(int status);
```
oh! I forgot a git add!!

nickdesaulniers replied:
fixed in f2ac0c382501

nickdesaulniers committed:
[libc][OSUtil] refactor quick_exit to be an object library everywhere

The usage of __builtin_unreachable after calls to quick_exit were distressing.
If a function is properly marked [[noreturn]] then __builtin_unreachable is not
necessary.

Looking into this further, we seem to have header only implementations for CPU
targets. The inline nature of these functions is curious; we're going to exit,
it doesn't matter if we need to pay the call of a function or not. If we just
make these functions have distinct TUs rather than be header only, we can clean
up the cmake rules for quick_exit which were different between CPU and GPU.

Remove darwin support for quick_exit. This isn't being tested, and we can bring
it back when necessary.

zeroomega commented:
Hi this patch breaks `runtimes-armv6m-unknown-eabi` build with error message:

```
[58/398](61) Building CXX object libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj
FAILED: libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj 
/b/s/w/ir/x/w/llvm_build/./bin/clang++ --target=armv6m-unknown-eabi -DLIBC_NAMESPACE=__llvm_libc_19_0_0_git -I/b/s/w/ir/x/w/llvm-llvm-project/libc -isystem /b/s/w/ir/x/w/llvm_build/include/armv6m-unknown-eabi --target=armv6m-unknown-eabi -mcpu=cortex-m0plus -mthumb -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -ffunction-sections -fdata-sections -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-armv6m-unknown-eabi-bins=../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O2 -g -DNDEBUG -std=gnu++17 -UNDEBUG --target=armv6m-unknown-eabi -fpie -ffreestanding -ffixed-point -fno-builtin -fno-exceptions -fno-lax-vector-conversions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -ftrivial-auto-var-init=pattern -Wall -Wextra -Werror -Wconversion -Wno-sign-conversion -Wimplicit-fallthrough -Wwrite-strings -Wextra-semi -Wnewline-eof -Wnonportable-system-include-path -Wstrict-prototypes -Wthread-safety -Wglobal-constructors -MD -MT libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj -MF libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj.d -o libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj -c /b/s/w/ir/x/w/llvm-llvm-project/libc/src/__support/OSUtil/baremetal/quick_exit.cpp
/b/s/w/ir/x/w/llvm-llvm-project/libc/src/__support/OSUtil/baremetal/quick_exit.cpp:12:1: error: an attribute list cannot appear here
   12 | [[noreturn]] extern "C" void __llvm_libc_quick_exit(int status);
      | ^~~~~~~~~~~~
1 error generated.
```
Failed build: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8752376881927918433/overview
Full stdout: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8752376881927918433/+/u/clang/build/stdout
Could you revert the change, fix it and reland it please?

jhuber6 commented:
> [[noreturn]] extern "C" void __llvm_libc_quick_exit(int status);

This should be trivial to fix-forward instead of revert, see https://godbolt.org/z/zv9oqnrd9.

jhuber6 commented:
> Hi this patch breaks `runtimes-armv6m-unknown-eabi` build with error message:
> 
> ```
> [58/398](61) Building CXX object libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj
> FAILED: libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj 
> /b/s/w/ir/x/w/llvm_build/./bin/clang++ --target=armv6m-unknown-eabi -DLIBC_NAMESPACE=__llvm_libc_19_0_0_git -I/b/s/w/ir/x/w/llvm-llvm-project/libc -isystem /b/s/w/ir/x/w/llvm_build/include/armv6m-unknown-eabi --target=armv6m-unknown-eabi -mcpu=cortex-m0plus -mthumb -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -ffunction-sections -fdata-sections -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-armv6m-unknown-eabi-bins=../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O2 -g -DNDEBUG -std=gnu++17 -UNDEBUG --target=armv6m-unknown-eabi -fpie -ffreestanding -ffixed-point -fno-builtin -fno-exceptions -fno-lax-vector-conversions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -ftrivial-auto-var-init=pattern -Wall -Wextra -Werror -Wconversion -Wno-sign-conversion -Wimplicit-fallthrough -Wwrite-strings -Wextra-semi -Wnewline-eof -Wnonportable-system-include-path -Wstrict-prototypes -Wthread-safety -Wglobal-constructors -MD -MT libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj -MF libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj.d -o libc/src/__support/OSUtil/baremetal/CMakeFiles/libc.src.__support.OSUtil.baremetal.baremetal_util.dir/quick_exit.cpp.obj -c /b/s/w/ir/x/w/llvm-llvm-project/libc/src/__support/OSUtil/baremetal/quick_exit.cpp
> /b/s/w/ir/x/w/llvm-llvm-project/libc/src/__support/OSUtil/baremetal/quick_exit.cpp:12:1: error: an attribute list cannot appear here
>    12 | [[noreturn]] extern "C" void __llvm_libc_quick_exit(int status);
>       | ^~~~~~~~~~~~
> 1 error generated.
> ```
> 
> Failed build: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8752376881927918433/overview Full stdout: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8752376881927918433/+/u/clang/build/stdout Could you revert the change, fix it and reland it please?

Let me know if https://github.com/llvm/llvm-project/commit/630283c38bce9ec188c96ddb90b8251afc2b3d8e fixes it.
